server_udp.go

```diff
@@ -42,83 +47,171 @@ func (w *UDPResponseWriter) WriteMsg(m *dns.Msg) error {
 		log.Tracef("Failed to encrypt the DNS query: %v", err)
 		return err
 	}
-
-	_, _ = udpWrite(res, w.udpConn, w.remoteAddr, w.localIP)
-	return nil
+	_, err = dns.WriteToSessionUDP(w.udpConn, res, w.sess)
+	return err
 }
 
-// ServeUDP - listens to UDP connections, queries are then processed by Server.Handler.
-// It blocks the calling goroutine and to stop it you need to close the listener.
+// ServeUDP listens to UDP connections, queries are then processed by Server.Handler.
+// It blocks the calling goroutine and to stop it you need to close the listener
+// or call Server.Shutdown.
 func (s *Server) ServeUDP(l *net.UDPConn) error {
-	// Check that server is properly configured
-	if !s.validate() {
-		return ErrServerConfig
-	}
-
-	// set UDP options to allow receiving OOB data
-	err := udpSetOptions(l)
+	err := s.prepareServeUDP(l)
 	if err != nil {
 		return err
 	}
 
-	// Buffer to read incoming messages
-	b := make([]byte, dns.MaxMsgSize)
+	// Tracks UDP handling goroutines
+	udpWg := &sync.WaitGroup{}
+	defer s.cleanUpUDP(udpWg, l)
+
+	// Track active goroutine
+	s.wg.Add(1)
+
+	log.Info("Entering DNSCrypt UDP listening loop on udp://%s", l.LocalAddr())
 
 	// Serialize the cert right away and prepare it to be sent to the client
-	certBuf, err := s.ResolverCert.Serialize()
+	certTxt, err := s.getCertTXT()
 	if err != nil {
 		return err
 	}
-	certTxt := packTxtString(certBuf)
 
-	// Init oobSize - it will be used later when reading and writing UDP messages
-	oobSize := udpGetOOBSize()
+	for s.isStarted() {
```

Comment:
I feel like the inside of this loop (and the one in the TCP one) should be a separate method, but I'll let you decide.